core/shared/src/main/scala/org/typelevel/log4cats/PagingSelfAwareStructuredLogger.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2018 Christopher Davenport
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.log4cats
+
+import cats._
+import cats.implicits._
+
+import java.util.UUID
+
+object PagingSelfAwareStructuredLogger {
+
+  /**
+   * Wrap a SelfAwareStructuredLogger adding pagination functionality.
+   *
+   * @param pageSizeK The size (unit is kilobyte) of the chunk of message in each page; this does not include the
+   *                  page header and footer, and tracing data.
+   * @param logger The SelfAwareStructuredLogger to be used to do the actual logging.
+   * @tparam F Effect type class.
+   * @return SelfAwareStructuredLogger with pagination.
+   */
+  def withPaging[F[_]: Monad](pageSizeK: Int, maxPageNeeded: Int)(
+      logger: SelfAwareStructuredLogger[F]
+  ): SelfAwareStructuredLogger[F] =
+    new PagingSelfAwareStructuredLogger[F](pageSizeK, maxPageNeeded)(logger)
+
+  private class PagingSelfAwareStructuredLogger[F[_]: Monad](pageSizeK: Int, maxPageNeeded: Int)(
+      sl: SelfAwareStructuredLogger[F]
+  ) extends SelfAwareStructuredLogger[F] {
+    private val pageIndices = 1 to maxPageNeeded
+
+    private def pagedLogging(
+        loggingLevelChk: => F[Boolean],
+        loggingOp: (=> String) => F[Unit],
+        msg: => String
+    ): F[Unit] = {
+      loggingLevelChk.ifM(
+        {
+          val pageSize = pageSizeK * 1000
+          val length = msg.getBytes().length
+          if (length <= pageSize)
+            loggingOp(msg)
+          else {
+            val decorationSize = 256
+            val pageContentSize = pageSize - decorationSize
+            val pageList = msg.grouped(pageContentSize).toList
```

Comment:
It's also going to allocate and hold all the substrings at once, instead of one substring at a time. This isn't normally a big deal, but if the log message is large enough to page, it might be large enough to matter.  Instead of `.grouped` out here, you might just calculate each chunk from the pageIndex inside the `traverse_` with a bit of `substring` math.

Reply:
Thanks for the insight, will just do that.